contracts/axelar-gateway/tests/test.rs

```diff
@@ -490,3 +492,346 @@ fn upgrade_unauthorized() {
     assert_invoke_auth_err!(not_owner, client.try_upgrade(&new_wasm_hash));
     assert_invoke_auth_err!(client.operator(), client.try_upgrade(&new_wasm_hash));
 }
+
+#[test]
+#[should_panic]
+/// TODO: figure out how to detect error in constructor failure
+fn fail_initialization_with_empty_signer_set() {
+    let env = Env::default();
+    env.mock_all_auths();
+    let owner = Address::generate(&env);
+    let operator = Address::generate(&env);
+
+    let empty_signer_set = Vec::<WeightedSigners>::new(&env);
+    let domain_separator: BytesN<32> = BytesN::random(&env);
+    let previous_signers_retention = randint(0, 10) as u64;
+    let minimum_rotation_delay = 0;
+    let initial_signers = empty_signer_set;
+
+    // should panic because of empty signer set
+    env.register(
+        AxelarGateway,
+        (
+            owner,
+            operator,
+            domain_separator,
+            minimum_rotation_delay,
+            previous_signers_retention as u64,
+            initial_signers,
+        ),
+    );
+}
+
+#[test]
+#[should_panic(expected = "failed ED25519 verification")]
+fn fail_validate_proof_invalid_signatures() {
+    let (env, signers, client) = setup_env(randint(0, 10), randint(1, 10));
+
+    let (message, _) = generate_test_message(&env);
+    let messages = vec![&env, message.clone()];
+
+    let msg_hash: BytesN<32> = BytesN::random(&env);
+    let proof = generate_proof(&env, msg_hash, signers);
+
+    // should panic, proof is for different message hash
+    // NOTE: panic occurs in std function, cannot handle explicitly
+    client.approve_messages(&messages, &proof);
+}
+
+#[test]
+fn fail_validate_proof_empty_signatures() {
+    let (env, signers, client) = setup_env(randint(0, 10), randint(1, 10));
+
+    let msg_hash: BytesN<32> = BytesN::random(&env);
+    let mut proof = generate_proof(&env, msg_hash.clone(), signers.clone());
+
+    // Modify signatures to make them invalid
+    let mut new_signers = Vec::new(&env);
+    for signer in proof.signers.iter() {
+        new_signers.push_back(ProofSigner {
+            signer: signer.signer,
+            signature: ProofSignature::Unsigned,
+        });
+    }
+    proof.signers = new_signers;
```

Comment:
```suggestion
    proof.signers.iter().for_each(|mut signer| signer.signature = ProofSignature::Unsigned);
```
maybe something like this instead?

Reply:
there isn't an `iter_mut` exposed so ultimately new `ProofSigner`s have to be created. I tried using `map` to do this functionally but I couldn't `collect()` into a `sorban_sdk::Vec` so I think the way it is acceptable for now